src/main/java/com/google/devtools/build/lib/analysis/CommandConstructor.java

```diff
@@ -0,0 +1,34 @@
+// Copyright 2019 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.devtools.build.lib.analysis;
+
+import com.google.common.collect.ImmutableList;
+import com.google.devtools.build.lib.actions.Artifact;
+
+/**
+ * The interface to construct command line for different shells (Bash, Batch, Powershell).
+ * Used in {@link com.google.devtools.build.lib.analysis.CommandHelper}
+ */
+public interface CommandConstructor {
+
+  // Given a string of command, return the arguments to run the command.
+  ImmutableList<String> buildCommandLineSimpleArgv(String command);
+
+  // Write the command to a script and return the artifact of the script.
+  Artifact buildCommandLineArtifact(RuleContext ruleContext, String command);
+
+  // Given an artifact of a script, return the arguments to run this command.
+  ImmutableList<String> buildCommandLineArgvWithArtifact(Artifact scriptFileArtifact);
```

Comment:
Since this function does the same as `buildCommandLineSimpleArgv` but with a script instead of a command, can we name them the same?

Reply:
Good idea!